### ISSUE TITLE
Support HW Intrinsics in crossgen under control of ISA options

### ISSUE THREAD

**Issue by 4creators** [dotnet/coreclr#20034]
Some applications using HW intrinsics could be short running tasks (i.e. image format conversion/resizing) which are launched by other applications on demand. In such scenarios it is beneficial to use R2R or native assembly to have faster startup time. Unfortunately time critical path will not be crossgened currently due to exclusion of methods containing HW intrinsics from compilation.

Crossgen is planned to be shipped as a public tool #19797 soon what will provide officially supported scenario to crossgen applications locally.

The main reason for excluding HW Intrinsics from R2R or native assemblies was to avoid ISA mismatches between architecture on which crossgen is running and target architecture on which native assembly will run.

My proposal is to introduce command line switches to `crossgen` which could be used to control compilation of methods containing HW Intrinsics globally by enabling selected ISAs like in native compilers.

Proposed design of command line options:

x86/x64 ISA

```
-Sse
-Sse2
-Sse3
-Ssse3
-Sse41
-Sse42
-Avx
-Avx2
-Bmi1
-Bmi2
-Fma
-Aes
-Pclm // for Pclmulqdq
-Popcnt
-Lzcnt
```

Arm64 ISA

```
-Aes      
-Atomics  
-Crc32    
-Dcpop    
-Dp       
-Fcma     
-Fp       
-Fp16     
-Jscvt    
-Lrcpc    
-Pmull    
-Sha1     
-Sha256   
-Sha512   
-Sha3     
-Simd     
-Simdv81 
-SimdFp16
-Sm3      
-Sm4      
-Sve
```

Depending on implementation some switches could be used to enable selected ISA and all lower ISAs. For instance `-Avx2` switch should enable all Sse switches and `-Avx`. In other cases switches have to be treated as an on/off command line switch.

It would be very helpful to use assembly level attribute (`AssemblySupportedIsa`) with data indicating against which ISAs assembly was compiled.

In future I would like to propose supporting method multiversioning for native/R2R assembly compilation with mechanism similar to used by GCC and libc26 and later.

**Comment by 4creators** [comment]
cc @CarolEidt @jkotas @AndyAyersMS @tannergooding @fiigii @eerhardt

**Comment by jkotas** [comment]
> time critical path will not be crossgened currently due to exclusion of methods containing HW intrinsics from compilation.

There are other classes of methods that are not precompiled today and that are more common than methods with hardware intrinsics. For example, PInvokes are not crossgened today.

If you want this to make a difference for startup, those more common cases need to be dealt with first.

**Comment by CarolEidt** [comment]
I think this would require a non-trivial amount of design and implementation, and would need to be justified by some compelling use cases, with measurements of how we would expect the throughput to be positively impacted. While such switches are useful and common in static compilers, those are scenarios where a JIT doesn't usually exist.

Such a design would have to be carefully thought through from a usability perspective:
- Is this per-assembly? That would seem to me to be less than ideal.
- If we were support multi-versioning in the longer term, how does this fit in?
- Would we define managed exceptions to be thrown for unsupported instructions?
- Would we want to support explicit multiversioning? (i.e. generate code for the `IsSupported` calls)

I'm sure this just scratches the surface of the design questions that would need to be addressed.

**Comment by 4creators** [comment]
> I'm sure this just scratches the surface of the design questions that would need to be addressed.

Agree, multiversioning is a very complex problem which requires complex design work. However, simpler design in a first step, where a particular ISAs could be activated for `crossgen`, provides very good solution for multiple scenarios. It could be for instance a deployment scenario where installed application is crossgened during installation (similar to how `ngen` works) or which selects assemblies to install which best support local ISAs.

> would need to be justified by some compelling use cases, with measurements of how we would expect the throughput to be positively impacted 

IMHO the scenarios where startup time is important are usually UI related and exist mainly in the desktop/mobile applications where developer intends to provide working UI with ready to use content as fast as possible after launching application. This may affect all graphics/media application where amount of optimized code which has to be jitted could be significant. The same is true for command line tools which are very often used in **professional** media/graphics processing workflow (i.e. encoders/decoders/transmuxers/data format converters/resizing/feature detection/segmentation etc.). Quite a lot of **medical image processing and viewing** apps are targeting .NET Framework and accompany high unit value diagnostic equipment, professional photo apps [Capture One Pro](https://www.phaseone.com/en/Products/Software/Capture-One-Pro/Highlights.aspx) from Phase One are written in C# and target .NET Framework/Mono and they accompany equipment with [unit worth of $100k and up](https://www.phaseone.com/en/Camera-Systems/XF-Camera-System/IQ4/XF-IQ4-150MP-Camera-System.aspx) used  worldwide by top world photographers.

Due to availability of HW intrinsics and WPF/WindowsForms support in  .NET Core 3.0 it will become attractive target for porting this class of applications.

The simplest scenario which I would expect to be often used is an app distribution packaged as an installer which contains multiple versions of crossgened `Codec.Assembly` targeting plain `x86`, `Sse41`, `Avx2` and in future `Avx512` ISAs. During installation installer through special action detects supported by target machine ISAs and selects for installation assembly with highest ISA version supported by machine. All this is supported currently by for instance WiX installer framework. 

Measurements are currently impossible due to classical egg chicken problem: applications on which we could measure throughput do not exist yet.
